Docs: Flow control section (#754)
inngest/website: pages/docs/guides/rate-limiting.mdx

export const description = 'Prevent excessive function runs over a given time period by skipping events beyond a specific limit. Ideal for protecting against abuse.'

# Rate limiting

Rate limiting is a _hard limit_ on how many function runs can start within a time period. Events that exceed the rate limit are _skipped_ and do not trigger functions to start. This prevents excessive function runs over a given time period. Some use cases for rate limiting include:

> * Preventing abuse of your system

Reviewer: ```suggestion
* Preventing abuse of your system.
```

> * Reducing frequency of data synchronization functions

Reviewer: ```suggestion
* Reducing frequency of data synchronization functions.
```

> * Smoothing out noisy [webhook](/docs/platform/webhooks) events

Reviewer: Not sure what "Smoothing out" means here. Would you have a common or Plain English expression to offer here?

## How to configure rate limiting

<CodeGroup forceTabs>
```ts
export default inngest.createFunction(
  {
    id: "synchronize-data",
    rateLimit: {
      limit: 1,
      period: "4h",
      key: "event.data.company_id",
    },
  },
  { event: "intercom/company.updated" },
  async ({ event, step }) => {
    // This function will be rate limited

>     // It will only run 1 once per 4 hours for a given event payload with matching company_id

Reviewer: ```suggestion
    // It will only run once per 4 hours for a given event payload with matching company_id
```

  }
);
```
</CodeGroup>

### Configuration reference

* `limit` - The maximum number of functions to run in the given time period.

> * `period` - The time period of which to set the limit. The period begins when the first matching event is received.

Reviewer: "The time period of which to set the limit." doesn't sound intuitive to me but I don't have a suggestion atm on how to make it easy.

> * `key` - An optional [expression](/docs/guides/writing-expressions) using event data to apply each limit too. Each unique value of the `key` has it's own limit, enabling you to rate limit function runs by any particular key, like a user id.

Reviewer: ```suggestion
* `key` - An optional [expression](/docs/guides/writing-expressions) using event data to apply each limit too. Each unique value of the `key` has its own limit, enabling you to rate limit function runs by any particular key, like a user id.
```

<Callout>

>   Any events received in excess of your `limit` are ignored. This means this is not the right approach if you need to process every single event sent to Inngest. Instead, check out [throttle](/docs/guides/throttling).

Reviewer: ```suggestion
  Any events received in excess of your `limit` are ignored. This means this is not the right approach if you need to process every single event sent to Inngest. Consider using [throttle](/docs/guides/throttling) instead.
```

</Callout>

## How rate limiting works

> Each time an event is received that match's your function's trigger, it is evaluated prior to executing your function. If `rateLimit` is configured, Inngest uses the `limit` and `period` options to only execute a maximum number of functions during that period.

Reviewer: ```suggestion
Each time an event is received that matches your function's trigger, it is evaluated prior to executing your function. If `rateLimit` is configured, Inngest uses the `limit` and `period` options to only execute a maximum number of functions during that period.
```

Inngest's rate limiting implementation uses the [“Generic Cell Rate Algorithm”](https://en.wikipedia.org/wiki/Generic_cell_rate_algorithm) (GCRA). To _overly simplify_ how this works, Inngest will use the `limit` and `period` options to create "buckets" of time in which your function can execute _once_.

```
limit / period = bucket time window
```

For example, this means that for a `limit: 10` and `period: '60m'` (60 minutes), the bucket time window will be 6 minutes. Any event triggering the function "fills up" the bucket for that time window and any additional events are ignored until the bucket's time window is reset. The algorithm (GCRA) is more sophisticated than this, but at the basic level - `rateLimit` ensures that you'll only run the max `limit` number of items over the `period` that you specify.

<Callout>
  Events that are ignored by the function will continue to be stored by Inngest.
</Callout>

> **How the rate limit is applied with a consistent rate of events received**

Reviewer: I love these graphics, thank you for creating them!

Two comments, though:
1. To some folks diagrams are not intuitive. Could you add a description of what is seen?
2. Less importantly, could you check if the reversed colors work here? If yes, we could easily add a dark-mode version (I have created a component like this a while ago, used it in Resend's guide I think)

Author: Good points! I think this is worth while, just may not be able to get it done properly for release this morning 😄

<img
  src="/assets/docs/reference/function-rate-limit.png"
  width="800"
  height={280}
  quality="95"
  alt="Visualization of how the rate limit is applied with a consistent rate of events received"
/>

**How the rate limit is applied with sporadic events received**

<img
  src="/assets/docs/reference/function-rate-limit-gaps.png"
  width="800"
  height={280}
  quality="95"
  alt="Visualization of how the rate limit is applied with sporadic events received"
/>

**How the rate limit is applied when limit is set to 1**

<img
  src="/assets/docs/reference/function-rate-limit-equal-1.png"
  width="800"
  height={280}
  quality="95"
  alt="Visualization of how the rate limit is applied when limit is set to 1"
/>

### Using a `key`

> When a `key` is added, a separate limit is applied for each unique value of the `key` expression. For example if your `key` is set to `event.data.customer_id`, each customer would have their individual rate limit applied to functions run meaning different users might have the same function run in same bucket time window, but two runs will not happen for the same `event.data.customer_id`. Read [our guide to writing expressions](/docs/guides/writing-expressions) for more info.

Reviewer: ```suggestion
When a `key` is added, a separate limit is applied for each unique value of the `key` expression. For example, if your `key` is set to `event.data.customer_id`, each customer would have their individual rate limit applied to functions run meaning different users might have the same function run in same bucket time window, but two runs will not happen for the same `event.data.customer_id`. Read [our guide to writing expressions](/docs/guides/writing-expressions) for more information.
```
